src/components/vm/vmDetailsPage.jsx

```diff
@@ -63,6 +65,10 @@ export const VmDetailsPage = ({
         // eslint-disable-next-line react-hooks/exhaustive-deps
     }, []);
 
+    // not initialized yet?
+    if (!vm.capabilities)
+        return <EmptyStatePanel title={ _("Loading...") } loading />;
+
```

Comment:
I don't think checking for vm.capabilities does what is indented here. 
If it's to handle only dummy VMs for the duration ofVM creation, then this should already be handled in [app.jsx](https://github.com/cockpit-project/cockpit-machines/blob/main/src/app.jsx#L297), where we show "Creating [VM]":
![Screenshot from 2023-04-12 10-12-06](https://user-images.githubusercontent.com/42733240/231395650-8e54d2de-5644-4aee-b3a7-edabd65c0809.png).

If on the other hand, this should fix some situation where we have object of a regular existing VM fetched through libvirt API, then checking for `vm.capabilities.` It's true it's a last property of VM which gets fetched through libvirt API, but as we can see [here](https://github.com/cockpit-project/cockpit-machines/blob/main/src/libvirtApi/domain.js#L712), vm object in our redux store doesn't get updated until every property about the VM is fetched.

So there are only 2 states:
1. There is no "vm" object in redux store, since the [domainGet](https://github.com/cockpit-project/cockpit-machines/blob/main/src/libvirtApi/domain.js#L654) hasn't finished yet. This is already handled by in [app.jsx](https://github.com/cockpit-project/cockpit-machines/blob/main/src/app.jsx#L150) with "Loading resources" spinner
2. there is "vm" object in redux store with every single property, including capabilities.

The point is, I think there should never be a point in time where there is a "vm" object without "capabilities" property.

Reply:
@martinpitt can you give me an example of a reproducer where you run into an app crash because vm.capabilities was not initialized yet?

Reply:
@skobyda : The reproducers are in the description: testCreateDownloadRhel is our #2 top flake, and it reproduces very well locally. (I used `TEST_BROWSER=firefox` on fedora-37 as that's what fails most often in CI).

> I think there should never be a point in time where there is a "vm" object without "capabilities" property

That would be very nice, but is unfortunately and demonstrably (as indicated by the oopses in the test) not true. I tried to describe it in the commit message. There are several places which put a stub VM into the store, e.g. `finishVmCreateInProgress()`, `usageStartPolling()`, `doUsagePolling()`.

I initially started with a [patch like this](https://paste.centos.org/view/4bca7b2e), but it's horribly error prone (as we'll keep forgetting this in the future). This seems much cleaner and more robust to me.

It would certainly be possible to rework all the places which do `store.dispatch(updateVm(props))` and `updateOrAddVm` etc. to not do the store update if the detection isn't finished yet. But I don't know the data model well enough to do that. My suggestion would be to keep this fallback to fix the oops, and remove it if/once you want to rework the data store that it meets this assumption?

Reply:
Resolved, see my comment at https://github.com/cockpit-project/cockpit-machines/pull/1015#pullrequestreview-1380908918